### PR TITLE
feat: add kepler-internal CRD

### PR DESCRIPTION
**Pull request by sthaha**
This commit adds a `KeplerInternal` CRD which is a internal API of Kepler Operator. Kepler CRD is now a facade for `KeplerInternal` Creating a `Kepler` now creates a corresponding `keplerinternal` CR with appropriate spec initialised. The KeplerInternal Controller responds creation of `kepler-internal` by deploying kepler.

This separation allows for providing stable API - Kepler or other higher level api (such as `powermon.openshift.io - PowerMonitoring`) which provides a limited set of stable functionalities to users.

In this PR, I have added the provision to set `export.deployment.Image`  in the `keplerinternal` API. 

**NOTE:*** KeplerInternal API can break at any point in time, so its usage is highly discouraged other than for testing experimental features.

### CODE REVIEW

**Comment by sthaha** [comment]
@vprashar2929 could  you please give this a go on openshift as well?

**Comment by sthaha** [comment]
Lets wait for @vprashar2929 to run the tests before merging this 👼

File: bundle/manifests/kepler.system.sustainable.computing.io_keplerinternals.yaml [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+          metadata:
+            type: object
+          spec:
+            description: KeplerInternalSpec defines the desired state of Kepler
+            properties:
+              exporter:
+                properties:
+                  deployment:
+                    properties:
+                      image:
```
Comment (vprashar2929) [inline bundle/manifests/kepler.system.sustainable.computing.io_keplerinternals.yaml:74]: Shouldn't we add a default image `quay.io/sustainable_computing_io/kepler:latest` for `keplerinternals` ?

**Comment by vprashar2929** [comment]
When kepler deployed using `Kepler` API the instance also visible in `KeplerInternal` as well
```
~ ❯ oc get kepler -o wide ; oc get keplerinternals -o wide                                                                                         ⎈ 
NAME     PORT   DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   AGE     NODE-SELECTOR                  TOLERATIONS
kepler   9103   6         6         6       6            6           2m29s   {"kubernetes.io/os":"linux"}   [{"operator":"Exists"}]
NAME     PORT   DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   AGE     IMAGE                                                   NODE-SELECTOR                  TOLERATIONS
kepler   9103   6         6         6       6            6           2m33s   quay.io/sustainable_computing_io/kepler:release-0.6.1   {"kubernetes.io/os":"linux"}   [{"operator":"Exists"}]
```

File: bundle/manifests/kepler.system.sustainable.computing.io_keplerinternals.yaml (hunk 1 above)
Comment (sthaha) [reply bundle/manifests/kepler.system.sustainable.computing.io_keplerinternals.yaml:74]: Yeah, we could do that but it does we may be running *any* image. My preference would be to let user decide.

**Comment by sthaha** [comment]
> When kepler deployed using Kepler API the instance also visible in KeplerInternal as well

Like the PR mentions, `Kepler` is now a facade over `KeplerInternal`, so when user creates `Kepler`, kepler-controller creates `KeplerInternal` and then `kepler-internal` controller creates the daemonset, etc ..

File: pkg/controllers/kepler_internal.go [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+		// retry since some error has occurred
+		logger.V(6).Info("Get Error ", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if kepler == nil {
+		// no kepler found , so stop here
+		logger.V(6).Info("Kepler Nil")
+		return ctrl.Result{}, nil
+	}
```
Comment (vimalk78) [inline pkg/controllers/kepler_internal.go:114]: nit: rename variable to `ki` . should the log say `KeplerInternal Nil`

File: pkg/controllers/kepler_internal.go [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+	logger.V(6).Info("Running sub reconcilers", "kepler", kepler.Spec)
```
Comment (vimalk78) [inline pkg/controllers/kepler_internal.go:116]: running kepler internal ?

Comment (vimalk78) [reply pkg/controllers/kepler_internal.go:116]: same comment for other logs in this file, should it say `kepler` or `keplerinternal` ?

File: pkg/controllers/kepler_test.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package controllers
```
Comment (vimalk78) [inline pkg/controllers/kepler_test.go:1]: add license comment header

File: pkg/controllers/kepler_internal.go [hunk 5]
```diff
@@ -0,0 +1,403 @@
+package controllers
```
Comment (vimalk78) [inline pkg/controllers/kepler_internal.go:1]: add license comment header

File: config/crd/bases/kepler.system.sustainable.computing.io_keplerinternals.yaml [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    controller-gen.kubebuilder.io/version: v0.12.1
+  name: keplerinternals.kepler.system.sustainable.computing.io
+spec:
+  group: kepler.system.sustainable.computing.io
+  names:
+    kind: KeplerInternal
+    listKind: KeplerInternalList
+    plural: keplerinternals
+    singular: keplerinternal
+  scope: Cluster
```
[the remark this answers is not in the document]
Comment (sthaha) [reply config/crd/bases/kepler.system.sustainable.computing.io_keplerinternals.yaml:15]: KeplerInternal act exactly like `Kepler`, it is cluster scoped but the `spec` contains a `namespace` where the deployment of `kepler` should happen.

File: pkg/controllers/kepler_test.go (hunk 4 above)
Comment (sthaha) [reply pkg/controllers/kepler_test.go:1]: +1

**Comment by sthaha** [comment]
@vimalk78 @sunya-ch , I am closing this PR in favour of https://github.com/sustainable-computing-io/kepler-operator/pull/309

**Pull request closed without merging**